serve.go

```diff
@@ -52,30 +54,42 @@ func Serve(handlers *types.FaaSHandlers, config *types.FaaSConfig) {
 		handlers.LogHandler = auth.DecorateWithBasicAuth(handlers.LogHandler, credentials)
 	}
 
+	hm := newHttpMetrics()
+
 	// System (auth) endpoints
-	r.HandleFunc("/system/functions", handlers.FunctionReader).Methods(http.MethodGet)
-	r.HandleFunc("/system/functions", handlers.DeployHandler).Methods(http.MethodPost)
-	r.HandleFunc("/system/functions", handlers.DeleteHandler).Methods(http.MethodDelete)
-	r.HandleFunc("/system/functions", handlers.UpdateHandler).Methods(http.MethodPut)
+	r.HandleFunc("/system/functions", hm.InstrumentHandler(handlers.FunctionReader, "")).Methods(http.MethodGet)
```

Comment:
If `hm.InstrumentHandler` returned an `http.Handler`, you could just wrap `r` once instead of on every single line like this. I would rather have an option to pass a name to skip metrics on the `metrics` vs the 30+ repetitions here. There is also prior art in the OTEL project for passing a `skipList` to the metrics wrapper to skip specific endpoints to avoid noise.

Reply:
But why can't you capture the path from the request? I guess that is the part that i missed earlier and the part that I also wasn't expecting. It just seems unnecessary compared to the implementations I have seen in things like the OTEL middleware, the various other open source middlewares for muxes like the [chi](https://github.com/766b/chi-prometheus/blob/master/middleware.go) and [gorilla/mux](https://github.com/labbsr0x/mux-monitor/blob/master/monitor.go), and OpenTracing, which have similar behaviors.

Reply:
Is this about aesthetics or a potential problem with the chosen functionality?

Reply:
it is primarily aesthetics and maintenance. I think there is enough prior art for applying middlewares at the mux level instead of applying them one by one to everything handler. Doing it one by one is more produce to developer error by forgetting.

Since it is aesthetics, ultimately it is your project your decision.

Reply:
Are you suggesting that a slice of paths to trim should be passed to a single handler instance?

How would you suggest the path rewriting works in that case? Separate "instances" are required to provide the path without a function in the URL.